Sources/Baggage/Baggage.swift

```diff
@@ -0,0 +1,237 @@
+//
+// See LICENSE.txt for license information
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+// ==== ----------------------------------------------------------------------------------------------------------------
+// MARK: Baggage
+
+/// A `Baggage` is a heterogeneous storage type with value semantics for keyed values in a type-safe
+/// fashion.
+///
+/// Its values are uniquely identified via `BaggageKey`s (by type identity). These keys also dictate the type of
+/// value allowed for a specific key-value pair through their associated type `Value`.
+///
+/// ## Defining keys and accessing values
+/// Baggage keys are defined as types, most commonly case-less enums (as no actual instances are actually required)
+/// which conform to the `Baggage.Key` protocol:
+///
+///     private enum TestIDKey: BaggageKey {
+///       typealias Value = String
+///     }
+///
+/// While defining a key, one should also immediately declare an extension on `BaggageProtocol`,
+/// to allow convenient and discoverable ways to interact with the baggage item, the extension should take the form of:
+///
+///     extension BaggageProtocol {
+///       var testID: TestIDKey.Value? {
+///         get {
+///           self[TestIDKey.self]
+///         } set {
+///           self[TestIDKey.self] = newValue
+///         }
+///       }
+///     }
+///
+/// For consistency, it is recommended to name key types with the `...Key` suffix (e.g. `SomethingKey`) and the property
+/// used to access a value identifier by such key the prefix of the key (e.g. `something`). Please also observe the usual
+/// Swift naming conventions, e.g. prefer `ID` to `Id` etc.
+///
+/// ## Usage
+/// Using a baggage container is fairly straight forward, as it boils down to using the prepared computed properties:
+///
+///     var context = Baggage.background
+///     // set a new value
+///     context.testID = "abc"
+///     // retrieve a stored value
+///     let testID = context.testID ?? "default"
+///     // remove a stored value
+///     context[TestIDKey.self] = nil
+///
+/// Note that normally a baggage should not be "created" ad-hoc by user code, but rather it should be passed to it from
+/// a runtime. For example, when working in an HTTP server framework, it is most likely that the baggage is already passed
+/// directly or indirectly (e.g. in a `BaggageContext` or `FrameworkContext`)
+///
+/// ### Accessing all values
+///
+/// The only way to access "all" values in a baggage context is by using the `forEachBaggageItem` function.
+/// The baggage container on purpose does not expose more functions to prevent abuse and treating it as too much of an
+/// arbitrary value smuggling container, but only make it convenient for tracing and instrumentation systems which need
+/// to access either specific or all items carried inside a baggage.
+public struct Baggage: BaggageProtocol {
```

Comment:
This type having the same name as its module will create issues: https://bugs.swift.org/browse/SR-898. One way I've worked around this in the past is to make the module name plural. So in this case, `import Baggages`.

Reply:
+1 on the issue @tanner0101 highlights wrt to module and main type name conflicts. personally not involve with the plurals tho

Reply:
I was struggling to come up with good module names here, open to ideas, the issues:

The main thing we want to be `Context` the type, the package therefore became `BaggageContext` since the context contains baggage.

We want a module that does not depend on logging because @lukasa and the team would prefer to not pull in logging to nio, so we need `Baggage` the type in a separate module -- how do we call that module? It would be more rarely used as most everyone, with some exceptions, would be using the `BaggageContext` module.

The package itself is called `swift-context` following `swift-log` etc, where the module then is `Logging`... we don't have a `-ing` verb we could use though.

WDYT could be a good module name for the Baggage?

Reply:
If it were me, I'd keep it simple and put everything in a single module `Tracing`. Call it SwiftTracing and be done with it. Then you'd have `Tracing.Context` and `Tracing.Baggage`. That's consistent with how SwiftLog and SwiftMetrics are structured. Real projects will pull in all of these dependencies anyway and I don't see why SwiftNIO shouldn't depend on SwiftLog. FWIW, early iterations of Vapor tried to perfectly optimize the dependency tree and it became a huge PITA that I've been trying to undo since. Don't create extra modules / packages that you don't _need_.

Reply:
> I don't see why SwiftNIO shouldn't depend on SwiftLog. 

That's an open question to @lukasa and co.

It's pretty much the only reason we do two modules here.